New settings for links

Bibo-Joshi opened the pull request:
Hi. After #12, I finally got around to play around with this more and would like to propose two additions:

<details>
  <summary>This part was moved to #14</summary>

1. Better support for nitpicky mode. Currently, `:paramref:` is resolved without checking if the resulting target exists. By checking against the index created in `build_index` within `lookup_params` and returning `None` if the target doesn't exist, we give sphinx a chance to emit a warning about an unresolved reference, if needed. This minimal-effort approach has two downsides: First, the `_sphinx_paramlinks_` prefix is still included in the warning, e.g.

    <pre>path\to\python-telegram-bot\telegram\ext\_basepersistence.py:docstring of telegram.ext._basepersistence.BasePersistence.insert_bot:1: WARNING: py:paramref reference target not found: _sphinx_paramlinks_bot</pre> 

    Secondly, I don't know how well it plays with the [`nitpick_ignore`](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-nitpick_ignore) and `nitpick_ignore_regex` settings (I guess the prefix could mean trouble here)
    Alternatively, one could add an event handler for the [`warn-missing-reference`](https://www.sphinx-doc.org/en/master/extdev/appapi.html?highlight=missing-reference#event-warn-missing-reference) event, but the [built-in logic](https://github.com/sphinx-doc/sphinx/blob/80b0a16e1c5f7266522a50284a003a0e17cf5ff7/sphinx/transforms/post_transforms/__init__.py#L170-L208) of that is actually somewhat elaborate. At least for me, adapting this for sphinx-paramlinks is not worth the effort - I can live with the prefix and so far I don't use the `nitpick_ignore` settings

</details>

---

2. Add options to a) turn off the insertion of the link symbol after the parameter name and b) to make the parameter name itself the link. This idea came up at https://github.com/python-telegram-bot/python-telegram-bot/pull/2798#discussion_r759343953. The reason is that with the link symbol only showing on hover, there is a large gap between parameter name & description/type. But showing the link symbol just all the time also doesn't look __too__ good:

    ![image](https://user-images.githubusercontent.com/22366557/144278517-5f991d49-ebf1-47fc-824d-342c39581f41.png)

    that built is currently available [here](https://python-telegram-bot.readthedocs.io/en/docs-referencing/telegram.bot.html#telegram.Bot.add_sticker_to_set.params.user_id)

If you like the general idea of those changes, I'm happy to discuss the implementation :) I can also split this PR up into two for the two different parts if you like that better.

Cheers!

zzzeek commented:
so there is "show the link symbol all the time", which ...is just a CSS change, right?   I dont personally like it that way, so I wouldn't want that CSS change by default, if that's how this works.  we could perhaps render a style class that refers to the two different approaches in the CSS.

then as far as "make the name the link", how does that look?  IIUC the build you have above is showing the link symbol still.   

I'm all for options here, so I think this should be just **one** configuruation parameter ``link_style``:

* `symbol_hover` - show a symbol on hover, the default
* `symbol_always` - show a symbol always
* `paramname` - make the paramname the link
* `None` - don't self-link the parameters

what do you think?

two PRs is easier for me to understand the two if it's not too much trouble

Bibo-Joshi commented:
> so there is "show the link symbol all the time", which ...is just a CSS change, right? I dont personally like it that way, so I wouldn't want that CSS change by default, if that's how this works. we could perhaps render a style class that refers to the two different approaches in the CSS.
>
> then as far as "make the name the link", how does that look? IIUC the build you have above is showing the link symbol still.

Sorry, I should have been more explicit. "show the symbol all the time" was my first approach before we had the idea of making the parameter name a link. I did with by manually overriding the CSS file and this PR doesn't change the hover behavior.

The change in the css file just makes sure that if the parameter name is a link, it doesn't change the text color. So except for the cursor changing on hovering the name, there is no visual indication for the link. I can try and provide you with an example built.

> I'm all for options here, so I think this should be just **one** configuruation parameter `link_style`:
> 
> * `symbol_hover` - show a symbol on hover, the default
> * `symbol_always` - show a symbol always
> * `paramname` - make the paramname the link
> * `None` - don't self-link the parameters
> 
> what do you think?

One parameter with different settings sounds reasonable 👍 I'll see if I implement the `symbol_always` option, as that's not a priority for me

> two PRs is easier for me to understand the two if it's not too much trouble

Sure, no worries. Will do that later tonight or in the next few days

Bibo-Joshi commented:
I've moved the nitpicky thing to #14 and implemented the new config scheme. I've also added some info to the docs about that and how to customize styling.
I've decided not to provide an option for hovering/not hovering the link symbol, since that can be customized by overriding the CSS file.

Also, [here](https://github.com/sqlalchemyorg/sphinx-paramlinks/files/7637021/paramname_link.zip) is a small preview of how it reneders with `paramlinks_style = 'param_name_and_link_symbol'`. The only change I made in the CSS here is to change `li > a.paramlink {` to `a.paramlink {` because for some reason the hovering doesn't work otherwise (probably a theme issue …)

zzzeek commented:
OK, ill try to have alook soon.

zzzeek commented:
you're still in my inbox.   just have a lot coming in

zzzeek commented on sphinx_paramlinks/sphinx_paramlinks.py:
```diff
@@ -148,7 +154,13 @@ class LinkParams(Transform):
+    link_style = LINK_SYMBOL
+
     def apply(self):
+        if self.link_style is None:
```
this would be clearer if it said "self.link_style is NO_LINK", i didnt at first see "NO_LINK" was None, probably that should be a string

zzzeek commented on sphinx_paramlinks/sphinx_paramlinks.py:
```diff
@@ -30,6 +30,12 @@ def _is_html(app):
 # https://www.sphinx-doc.org/en/master/extdev/deprecated.html
 
+# Constants for link styles
+NO_LINK = None
```
can this be "no_link"?  the None value is easy to occur for other broken reasons

zzzeek commented on sphinx_paramlinks/sphinx_paramlinks.py:
```diff
@@ -197,19 +209,34 @@ def apply(self):
-                            classes=["paramlink", "headerlink"],
-                        ),
-                    )
+                    if self.link_style in [PARAM_NAME, PARAM_NAME_AND_LINK_SYMBOL]:
```
this is cool.  if you are pushing another change, change the list `[]` to tuple `()`, `link_style in (PARAM_NAME, PARAM_NAME_AND_LINK_SYMBOL)` , hardly makes a difference here but is more conventional

zzzeek reviewed (changes requested):
happy new year!   im back.   i think this is good, tiny stylistic changes that are not super essential but here we are

Bibo-Joshi commented:
Happy new year to you as well :) The requested changes make sense, so I pushed some along with adapting the readme.

zzzeek commented:
OK this looks great.  going to squash the commits and reformat README and that's it!

zzzeek commented:
OK I'm going to force-push some changes here, but also param_name_and_link_symbol is nesting the <a> tags, will see if I can fix that

```
<a class="paramname reference internal" href="#sqlalchemy.types.TypeDecorator.process_bind_param.params.dialect"><strong>dialect</strong><a class="paramlink headerlink reference internal" href="#sqlalchemy.types.TypeDecorator.process_bind_param.params.dialect">¶</a></a>
```

zzzeek commented:
there is also a really weird and subtle recursion overflow happening here if i touch that code at all, i have a vague idea why it's happening but it's really weird.   fixing that

zzzeek commented:
OK note major differences, if you can re-review please:

* new names for the parameter and the values
* use of Enum
* link_style is avaiable from .app, no need for the global
* changing the reference logic a bit to avoid nested <a> tags, recursion overflow

I've tested a bunch and it seems to work for me, if you can give it a run to make sure it's doing what you need as well, thanks.

Bibo-Joshi commented:
Ah, I hadn't noticed the nesting thing. Thanks for finding! Enum & `.app` make it cleaner as well 👍 
I added a more user friendly error message in case of a wrong config value and adapted the readme a bit, since plain rst doesn't know the `.. versionchanged::` directive of sphinx - GitHub displays it as

![image](https://user-images.githubusercontent.com/22366557/147870450-dab7cc8a-16bd-41f5-8eb8-ddc5d87e4cd6.png)

One comment on `HyperlinkStyle.BOTH`: If a third option is added at some point, `'both'` doesn't really make sense anymore, that's why I had chosen `LINK_SYMBOL_AND_NAME`. Anyway, that's probably rather unlikely …

zzzeek commented:
yes i thought if another symbol is added, BOTH starts to not make sense.   will think about another name, but it starts to look like maybe the parameter should itself be a tuple, but that's probably not worth it.   "versionadded" was totally a mistake as I thought i was in sphinx :) .

zzzeek commented:
OK please give my latest commit one more review and we can merge.

Bibo-Joshi commented:
LGTM 👌

zzzeek commented:
great!  thanks. will try to look at your next one in the coming week.
